File: day7/day7.py
import functools

commands = ['NOT', 'AND', 'OR', 'LSHIFT', 'RSHIFT']
data = open('day7/day7.txt').read().split('\n')
data = [d.split(' -> ') for d in data]
data = {d[1]:d[0] for d in data}
seen = {}

OPERATORS = {
    None: lambda arg: int(arg[0]) if arg[0].isdigit() else arg[0],
    "NOT": lambda arg: ~int(arg[1]),
    "AND": lambda arg: int(arg[0]) & int(arg[2]),
    "OR": lambda arg: int(arg[0]) | int(arg[2]),
    "LSHIFT": lambda arg: int(arg[0]) << int(arg[2]),
    "RSHIFT": lambda arg: int(arg[0]) >> int(arg[2]),
}

def get_wire(wire):
    try:
        return int(wire)
    except:
        pass
    cmd = data[wire].split(' ')
    if len(cmd) == 1 and cmd[0] in seen:
        seen[wire] = seen[cmd[0]]
    else:
        if len(cmd) == 1:
            seen[wire] = get_wire(OPERATORS[None](cmd))
        elif 'NOT' in cmd:
            cmd[1] = seen.get(cmd[1], get_wire(cmd[1]))
            seen[wire] = OPERATORS['NOT'](cmd)
        else:
            cmd[0] = seen[cmd[0]] if cmd[0] in seen else get_wire(cmd[0])
            cmd[2] = seen[cmd[2]] if cmd[2] in seen else get_wire(cmd[2])
            seen[wire] = OPERATORS[cmd[1]](cmd)
    return seen[wire]

p1 = get_wire('a')
seen = {}
data['b'] = str(p1)
p2 = get_wire('a')

print(f'Part 1: {p1}')
print(f'Part 2: {p2}')

File: day7/day7.js
const fs = require('fs');
const input = fs.readFileSync("day7/day7.txt", 'utf-8').split('\n');
var data = input.map(line => line.split(' -> ')).reduce((acc, line) => {
    acc[line[1]] = line[0];
    return acc;
}, {});

var seen = {};

const operators = {
    NOT: (arg) => ~arg[1],
    AND: (arg) => arg[0] & arg[2],
    OR: (arg) => arg[0] | arg[2],
    LSHIFT: (arg) => arg[0] << arg[2],
    RSHIFT: (arg) => arg[0] >> arg[2]
}

function getWire(wire) {
    if (!isNaN(wire)) return +wire;
    if (wire in seen) return seen[wire];
    var cmd = data[wire].split(" ");
    var res = 0;
    if (cmd.length == 1) return getWire(cmd[0]);
    if (cmd.includes("NOT")) {
        cmd[1] = getWire(cmd[1]);
        res = operators["NOT"](cmd);
    }
    else {
        cmd[0] = getWire(cmd[0]);
        cmd[2] = getWire(cmd[2]);
        res = operators[cmd[1]](cmd);
    }
    seen[wire] = res;
    return res;
}

function part1() {
    return getWire('a');
}

function part2() {
    seen = {b: part1()};
    return getWire('a');
}

console.log("Part 1: " + part1());
console.log("Part 2: " + part2());